### ISSUE TITLE
[HOLD for payment 2023-11-09] [$1000] Web – DOB – Error appears when year is selected using keyboard but doesn’t trough mouse.

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#24198]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1.	Open your user profile.
2.	Click on the "Settings" option.
3.	Select "Personal Details."
4.	Choose "Date of Birth."
5.	Pick a specific "Year."
6.	Use the mouse to select any year by scrolling, and observe that no error occurs.
7.	Go back to the previous screen.
8.	Once again, click on "Date of Birth."
9.	Select the desired "Year."
10.	This time, use the keyboard (Up or Down arrow keys to scroll) to choose a year, and then press the "Enter" key.
11.	Notice that an error message appears, stating "Please select a valid date."

## Expected Result:
Selecting year using keyboard closes year picker modal but should not submit the form automatically

## Actual Result:
An error message is shown when a year is selected using the keyboard, whereas no error message appears when the year is selected using the mouse.

## Workaround:
unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [X] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.49-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://github.com/Expensify/App/assets/43995119/45b4d484-ad13-429f-98e6-0f72c0232cb9

https://github.com/Expensify/App/assets/43995119/e33f1a6e-f92e-4161-a0e7-f8760848c818

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Usmantariq96
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690950105791819

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01141507e0c0672ac6</li>
        <li>Upwork Job ID: 1690361323494866944</li>
        <li>Last Price Increase: 2023-09-02</li>
    </ul>
</details>

**Comment by mvtglobally** [comment]
@sonialiap  Can you please also double check https://expensify.slack.com/archives/C049HHMV9SM/p1690955359134649 
This seems similar issue. Let em know if it should be logged separately or has a same root cause

**Comment by mvtglobally** [comment]
@hungvu193 Can you please share the link for the dup. I wasnt able to find the GH issue

**Comment by lukemorawski** [comment]
hey. It's Lucas from Callstack.
I think this should be fixed. By no means this is a desirable behaviour. The user selects the year by pressing Enter and doesn't want to submit the whole birthdate form.
Happy to take a look a this.

**Comment by lukemorawski** [comment]
I'm making an internal proposal today 🤞

**Comment by lukemorawski** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When a user wants to change the DOB in the Settings > Profile > Personal Details > Date of birth, opens the Year selection modal and selects the year by pressing `Enter`, that creates a validation error on the Date of birth form.
### What is the root cause of that problem?
The `<FormSubmit />` component's submit action is done by using `onKeyDownCapture` synthetic keyboard event callback prop on RNW `<View />` component that encapsulates the form.
When the app shows the year modal, it registers a keyboard shortcut for `Enter` key using `useKeyboardShortcut` hook. This is approach is utilised in couple of other places in the app. When user presses Enter on the year selection modal it not only selects the year, but also fires the FormSubmit `onSubmit` callback, triggering the validation. Since the user might had not selected the full date yet, the form shows a validation error.
### What changes do you think we should make in order to solve the problem?
In order to avoid triggering validation in forms when displaying a modal that registers Enter key as shortcut, the form should detect if a keyboard shortcut of that kind is currently registered and disable submitting until the shortcut is unregistered (upon modal unmount).
This can be done by implementing a new function `isShortcutRegistered` in the `KeyboardShortcut` module and invoking it in the `onSubmit` handler in the `<FormSubmit />` component when a user presses Enter. If the shortcut for enter is registered and user pressed enter then just do an early return and don't call the `onSubmit` callback.
### What alternative solutions did you explore?
None.

**Comment by lukemorawski** [comment]
@sonialiap any updates on this?

**Comment by sonialiap** [comment]
In my test I got a behavior similar to the issue in https://expensify.slack.com/archives/C049HHMV9SM/p1690955359134649

I do not get the validation error, but my selected year is not saved (mouse click or keyboard) if I don't also re-select the month and day:

https://github.com/Expensify/App/assets/17131195/26d9da34-6404-4562-b1f9-799248a15dad

A bug we should fix imo

**Comment by devrari** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
An error message is shown when a year is selected using the keyboard  while updating date of birth , whereas no error message appears when the year is selected using the mouse.
### What is the root cause of that problem?

We are not stopping the  propagation  of enter key from year select modal to  the  form submit. 

### What changes do you think we should make in order to solve the problem?
I am proposing following changes to fix  the issue  .  
1.  Add a  new  proptype in  selectionproptypes and   make it default as  false.
    ```
        /** Stop propagation */
         stopPropagation : PropTypes.bool,
 
2.   Pass the key to  config  in useKeyboardShortcut . 
```
         useKeyboardShortcut(
        CONST.KEYBOARD_SHORTCUTS.ENTER,
        () => {
            const focusedOption = flattenedSections.allOptions[focusedIndex];
            if (!focusedOption) {
                return;
            }
            props.onSelectRow(focusedOption);
        },
            {
            captureOnInputs: true,
            stopPropagation: props.stopPropagation,
            shouldBubble: () => !flattenedSections.allOptions[focusedIndex],
            },
        );
```    

3.  update useKeyboardShortcut.js to include  the  changes for config.
4.  Add code to  stop propagation in  bindHandlerToKeyDownEvent/index.js.
```
   if (callback.stopPropagation) {
            event.stopPropagation();
   }
```
5.  Pass  stopPropagation from YearPickerModal.js 
 ```
       <SelectionListRadio
                shouldDelayFocus
                stopPropagation
                textInputLabel={translate('yearPickerPage.selectYear')}
                textInputValue={searchText}
                textInputMaxLength={4}
                onChangeText={(text) => setSearchText(text.replace(CONST.REGEX.NON_NUMERIC, '').trim())}
                keyboardType={CONST.KEYBOARD_TYPE.NUMBER_PAD}
                headerMessage={headerMessage}
                sections={sections}
                onSelectRow={(option) => props.onYearChange(option.value)}
                initiallyFocusedOptionKey={props.currentYear.toString()}
            />
```
### What alternative solutions did you explore?
None

**Comment by lukemorawski** [comment]
@sonialiap Proposal by @devrari is a more elegant and streamlined solution! Great job! 
However the bug you have showed is not caused by the solution to the year selection/validation problem. It's present even before the changes. There should be a separate ticket for that IMO.

**Comment by devrari** [comment]
@sonialiap @aimane-chnaif Hi. I am new to contributing to expensify. Please let me know if I need to do anything else to get this task assigned to me?
Thanks.

**Comment by devrari** [comment]
@aimane-chnaif  when we open YearPickerModal and we use  arrow  key and enter key to select year at  that point of  time, 
we  uses useKeyboardShortcut hook  to perform the action which if not  stopped will bubble up till the form submission.
when  we are not  opening the modal, its  just date select  that  is being  called and onsubmit  is not called  automatically

**Comment by devrari** [comment]
@aimane-chnaif but as far as I can tell its  because of useKeyboardShortcut hook only form get  submitted automatically  in 
 case of enter key event. 


> I am not convinced. Form submit button is on Date of birth page, not modal.
The modal  is inside the form, so enter  key  submission  propogates till the form  submission.

**Comment by devrari** [comment]
@joh42 there are places , where after pressing enter form submission is the desired behaviour, because of which I had proposed to introduce a flag based on which we will stop propagation.

**Comment by joh42** [comment]
I do not actually think there are scenarios where we would want the enter event to propagate beyond the BaseSelectionListRadio component. This would lead to inconsistent behavior depending on whether the user selected an option using their mouse or using their keyboard

IE if a user selected an option using their keyboard, there may be validation triggered in ancestral components, like the form validation encountered in this bug. But that would not happen if a user selected an option using their mouse, so we cannot rely on that validation being triggered.

In my opinion, we should get the same behavior regardless of how the user selected an option, but I could be wrong.

**Comment by devrari** [comment]
We are using submission on enter key press while selecting country and that is desired behaviour. 
And as far as the validation is getting triggered. Validation will get triggered in both the cases when form gets submitted.

**Comment by devrari** [comment]
@aimane-chnaif  with the latest  expensify app code  changes I can see  that BaseSelectionListRadio.js is no longer  present  and it  is  replaced by BaseSelectionList.js, 
changes to fix the issue however remains the  same , just instead  of  in BaseSelectionListRadio.js  we have  make  changes to BaseSelectionList.js.

**Comment by joh42** [comment]
@aimane-chnaif The reason form validation is only triggered with the [YearPickerModal](https://github.com/Expensify/App/blob/main/src/components/NewDatePicker/CalendarPicker/YearPickerModal.js) is opened is that the enter event is fired from the "Please select a year" input field. If we blur that input field, form validation will not be triggered when we press enter despite the [YearPickerModal](https://github.com/Expensify/App/blob/main/src/components/NewDatePicker/CalendarPicker/YearPickerModal.js) being open. In the video below, I first press enter with the input field blurred, which does not trigger validation. Then I press enter with the input field focused, which does trigger validation:

https://github.com/Expensify/App/assets/138504087/39b1c4ea-58bb-4b45-8aa1-b7d72018785d


I once again want to make the case for fixing this everywhere in the app at once. Currently, the behavior when using a keyboard is not consistent with the behavior when using a mouse:

https://github.com/Expensify/App/assets/138504087/e28e5152-0ad5-46da-be43-3e79688fbe56


As such, I suggest we do not lock consistent behavior behind a prop, as that would lead to us having to pass that prop everywhere the [SelectionList](https://github.com/Expensify/App/blob/main/src/components/SelectionList/BaseSelectionList.js) is being used.

Rather, I suggest we stop the propagation of all enter events in the [SelectionList](https://github.com/Expensify/App/blob/main/src/components/SelectionList/BaseSelectionList.js) as proposed in [my proposal above](https://github.com/Expensify/App/issues/24198#issuecomment-1686553864). That will only require a 4-line change in one file.

It's worth noting that there was [another issue](https://github.com/Expensify/App/issues/23834) where the BZ member wrote that submitting when an option has been selected was the expected behavior. No one in that issue pointed out the inconsistency between using mouse/keyboard however. I'm going to leave a comment on that issue pointing this out. If we want to submit when an option has been selected, we should do it by passing a callback to the onSelectRow prop, as that will lead to consistent behavior across mouse/keyboard.

**Comment by devrari** [comment]
> The bug to fix here is:
> 
> > Date of birth page closes on Enter key press from keyboard, while doesn't through mouse
> 
>  Screen.Recording.2023-08-29.at.9.56.53.AM.mov 
> @sonialiap can you please update OP if you agree?

@aimane-chnaif In your case you are not seeing error because you already have a DOB  stored in your profile, and hence when form get  submitted your old date gets submitted and you don't get  validation error.
In  case a user  don't have his dob  set he will get the validation error in case of Enter key press from keyboard.

**Comment by devrari** [comment]
@aimane-chnaif correct, basically what I am saying is root cause is the same for both the cases, if dob is set you will see the dob page close , if dob is not set you will see the validation error.

**Comment by joh42** [comment]
> The bug to fix here is:
> 
> > Date of birth page closes on Enter key press from keyboard, while doesn't through mouse
> 
>  Screen.Recording.2023-08-29.at.9.56.53.AM.mov 
> @sonialiap can you please update OP if you agree?

Specifically, if you have selected a date of birth already, the DOB page will be closed, otherwise validation will run and show errors on the DOB page

**Comment by joh42** [comment]
> @devrari if the root cause is fixed, your error will be gone

But correct, if the root cause is fixed, no errors will show because validation will not be triggered

**Comment by devrari** [comment]
> Updated to
> 
> > ## Expected Result:
> > If you have selected a date of birth before running this test, then the DOB page will close with the selected date, without error. Otherwise, if no year is selected, validation will run and show errors on the DOB page.

@sonialiap  what you have updated in the expected result , that is the current behaviour  as it is. 
Can you please let me know what exactly  is the expected behaviour.

**Comment by devrari** [comment]
> Oops, I think this is Actual Result which should be fixed.

@aimane-chnaif, In both the cases whether DOB is set  already or  not the expected behaviour  should be,  to select the  year and don't fire the validation right  away  ?

**Comment by devrari** [comment]
@aimane-chnaif are you expecting anything else from my side before finalising on the proposal ? My existing proposal still works.

**Comment by joh42** [comment]
@sonialiap @aimane-chnaif Isn't the expected behavior that you get sent back to the date picker? That's how it currently works when you select an option using the mouse. As opposed to the DOB page closing entirely

**Comment by joh42** [comment]
Just FYI, I haven't heard anything on the [issue](https://github.com/Expensify/App/issues/23834) regarding the country page. Given that the behavior isn't consistent across mouse/keyboard, I still suggest we fix this everywhere at once.

**Comment by joh42** [comment]
@aimane-chnaif can you please review my proposal? It's been ready for 2 weeks

**Comment by joh42** [comment]
@aimane-chnaif The reason form validation is only triggered with the [YearPickerModal](https://github.com/Expensify/App/blob/main/src/components/NewDatePicker/CalendarPicker/YearPickerModal.js) is opened is that the enter event is fired from the "Please select a year" input field. If we blur that input field, form validation will not be triggered when we press enter despite the [YearPickerModal](https://github.com/Expensify/App/blob/main/src/components/NewDatePicker/CalendarPicker/YearPickerModal.js) being open. In the video below, I first press enter with the input field blurred, which does not trigger validation. Then I press enter with the input field focused, which does trigger validation:

https://github.com/Expensify/App/assets/138504087/39b1c4ea-58bb-4b45-8aa1-b7d72018785d

TLDR: only enter events from textinputs trigger form onSubmit

**Comment by joh42** [comment]
Also updated my proposal to explain the root cause in more detail

**Comment by devrari** [comment]
> @devrari please explain details about your root cause. I noticed onSubmit on Form is called only when YearPickerModal is opened. Can you find out why it's not called when modal is not opened?

@aimane-chnaif YearPickerModal is part of dateofbirth page, and it is inside the Form itself, and  hence when we use enter key the hook useKeyboardShortcut is called and it gets bubbled till  form  submission  and triggers the validation which is  failing , where as when we don't use keyboard to open the modal , the hook is not called  and hence we did not see the  validation error.

**Comment by sonialiap** [comment]
@aimane-chnaif where are we on this one?

**Comment by sonialiap** [comment]
@aimane-chnaif does that mean that the issue reported here is expected?

**Comment by devrari** [comment]
> ## Expected Result:
> If you have selected a date of birth before running this test, then the DOB page will close with the selected date, without error. Otherwise, if no year is selected, validation will run and show errors on the DOB page.

@sonialiap  this is the actual Result right now .  
I think the expected is 
To close  YearpickerModal once we select the year and don't submit the  form automatically. 
@aimane-chnaif  correct me if I am  wrong.

**Comment by sonialiap** [comment]
"Expected Result" in OP updated to 
>Selecting year using keyboard closes year picker modal but should not submit the form automatically

**Comment by sonialiap** [comment]
I believe we need an internal engineer on this. The label was added Aug 28 but no one got assigned, going to try it again https://github.com/Expensify/App/issues/24198#event-10207225827

**Comment by devrari** [comment]
@sonialiap just a reminder I do have a proposal that solves the issue  :)

**Comment by joh42** [comment]
Also to state it once more, onSubmit is only called when clicking a button or pressing enter while editing a field inside of a form, which is why this issue only exists when the year input field is focused:

> The submit event fires when the user clicks a [submit button](https://developer.mozilla.org/en-US/docs/Glossary/Submit_button) or presses Enter while editing a field (e.g. [<input type="text">](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/text)) in a form.

From https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event

**Comment by sonialiap** [comment]
@aimane-chnaif looks like there's a comment waiting for your reply: https://github.com/Expensify/App/issues/24198#issuecomment-1742613001

https://github.com/Expensify/App/issues/24198#issuecomment-1742416059, did you get a chance to review @devrari's proposal?

**Comment by joh42** [comment]
(and mine 😄)

**Comment by joh42** [comment]
@sonialiap maybe @aimane-chnaif isn't available?

**Comment by Li357** [comment]
@aimane-chnaif any updates?

**Comment by sonialiap** [comment]
@aimane-chnaif gentle bump :)

**Comment by sonialiap** [comment]
P.S. I'm OOO until Oct 22 so I won't be around to push the issue forward until then 👋

**Comment by sonialiap** [comment]
I'm OOO Oct 16-23. Adding leave buddy

Status: @aimane-chnaif was planning on making an update on this issue today

**Comment by kadiealexander** [comment]
@aimane-chnaif any update?

**Comment by devrari** [comment]
@aimane-chnaif thanks. 
I will compile a list of places [ BaseSelectionList ](https://github.com/Expensify/App/blob/main/src/components/SelectionList/BaseSelectionList.js) is being used.

**Comment by devrari** [comment]
@aimane-chnaif  this issue is still marked as internal ?

**Comment by kadiealexander** [comment]
@Li357 bump!

**Comment by sonialiap** [comment]
Thanks @kadiealexander for keeping the ball rolling! I'm back from OOO can take this back. Feel free to keep yourself assigned or remove yourself

**Comment by sonialiap** [comment]
@Li357 looks like we're waiting for you to approve the proposal. Aimane is 👍 https://github.com/Expensify/App/issues/24198#issuecomment-1765826610, what do you think?

**Comment by Li357** [comment]
The proposal looks good, assigning you @devrari. Let's make sure we explain why we need to stop propagation with some comments

**Comment by devrari** [comment]
@aimane-chnaif I have submitted the PR https://github.com/Expensify/App/pull/30300 . Please review and let me know.
Thanks

**Comment by devrari** [comment]
@aimane-chnaif  A humble request if you can please review the PR? . It will really help to meet  72 hours deadline. 
thanks.

**Comment by devrari** [comment]
@Li357 can you please review the PR?

**Comment by devrari** [comment]
@Li357 @aimane-chnaif  I  have updated the code by merging the latest changes from  expensify app as they have added typescript support for keyboardshortcut, and  as result of that keyboardshortcut/index.js was  deleted and index.ts  was added .

**Comment by devrari** [comment]
@aimane-chnaif @sonialiap  I see the issue is still marked as internal and the Upwork job  corresponding to this issue is no longer available. This is my first contribution so just curious what  is the next course of action here.

**Comment by sonialiap** [comment]
Posted new upwork job https://www.upwork.com/jobs/~01a16e55df7cf20b04

**Comment by sonialiap** [comment]
@devrari fix + bonus = $1500 - I'm not finding you on upwork, could you please apply to the job [here](https://www.upwork.com/jobs/~01a16e55df7cf20b04)? 
@aimane-chnaif review + bonus = $1500 - offer sent
@usmantariq96 report = $250 - offer sent

**Comment by usmantariq96** [comment]
@sonialiap offer accepted
Thanks

**Comment by devrari** [comment]
@sonialiap  I have  applied for the job  on Upwork . 
Thanks

**Comment by devrari** [comment]
> @sonialiap I have applied for the job on Upwork . Thanks

@sonialiap , just a reminder  :)

**Comment by sonialiap** [comment]
@devrari - hired and offer sent. The bonus will be added when the payment is made :) 

aimane-chnaif - paid
usmantariq96 - paid

**Comment by devrari** [comment]
@sonialiap - I have accepted the offer.

**Comment by sonialiap** [comment]
@devrari - paid ✔️

**Comment by sonialiap** [comment]
@aimane-chnaif bump for the checklist :)
